### ISSUE TITLE
Enabled is a special meaning term that requires a definition

### ISSUE THREAD

**Issue by dholmes-epcc-ed-ac-uk** [mpi-forum/mpi-issues#421]
# Problem

Enabled a special-meaning term that requires a formal definition.

# Proposal

chap-pt2pt "all enabled operations"
chap-dynamic "communication is enabled and can proceed"
chap-one-side-2 "communication is enabled"

Should be distinguished from
chap-tools "profiling is enabled"
chap-io-2 "atomic mode [is] enabled"
chap-one-side-2 "access to the windows is enabled"

# Changes to the Text

Describe the text changes here.

# Impact on Implementations

Describe changes that implementations will be required to make here.

# Impact on Users

Describe the changes that will impact users here.

# References

PR: https://github.com/mpi-forum/mpi-standard/pull/760

**Comment by Wee-Free-Scot** [comment]
Slides from a virtual meeting on this topic presented on 12th October 2022
[Enabled is an MPI term.pdf](https://github.com/mpi-forum/mpi-issues/files/9823172/Enabled.is.an.MPI.term.pdf)

Feedback from that presentation:
* Defining enabled as a term is a good direction
* Link with MPI_PREPARE -- an enabled term will make text for prepare easier -- potentially should be merged
* Link with holistic progress -- an enabled term will make text for progress easier -- potential dependence/interaction
* Therefore, this should be targeting MPI-4.1

**Comment by Wee-Free-Scot** [comment]
PDF showing proposed changes:

[mpi41-report_Issue421_Enabled_is_a_semantic_term.pdf](https://github.com/mpi-forum/mpi-issues/files/10056442/mpi41-report_Issue421_Enabled_is_a_semantic_term.pdf)

**Comment by Wee-Free-Scot** [comment]
Latest PDF, for reading at the Dec 2022 meeting:

[mpi41-report_Issue421_Enabled_is_a_semantic_term.pdf](https://github.com/mpi-forum/mpi-issues/files/10154778/mpi41-report_Issue421_Enabled_is_a_semantic_term.pdf)

See Section 2.4.1, page 12, lines 15-23.

**Comment by jprotze** [comment]
What are semantically-related procedures/procedure calls?
When is a happens-before relationship satisfied?
I think, the definition lacks a reference to the MPI operation. According to the current wording, all HB relationships of the whole program must be satisfied. 

The problem with the HB-relation is that it is usually defined for a completed program execution. Sometimes it is also defined for the execution up to a certain point (i.e., the current state of execution). I think, what you are trying to express is:
There is no outstanding MPI procedure call that would be in HB relation and is semantically-related.

Probably change the definition to something along:

Enabled: All semantically-related MPI procedure calls at all MPI processes in happens-before relation with the MPI operation were called - ... .


Is it necessary that all MPI procedure calls that are in HB-relation are called, or only calls ordered before the MPI operation according to the HB-relation?

As a small example: Rank 0 calls Isend and Wait. Rank 1 calls a matching Irecv and Wait. The HB relation ("of semantically-related MPI procedure calls"?) consists of: 
Isend@0 -> Wait@0
Irecv@1 -> Wait@1
Isend@0 -> Wait@1

Does it mean, that Wait@1 enables Isend@0, i.e. Isend@0 is not enabled unless Wait@1 is called?

**Comment by Wee-Free-Scot** [comment]
The receive operation is enabled by "Isend@0". The data cannot be received without the send operation being started.

The send operation is enabled by "Irecv@1". The data might not be transmitted until the receive operation has been started (depending on implementation-dependent buffer space availability).

The "Wait@1" for the receive operation will delay its return until the data has been written into the receive buffer, which requires the send operation to have been started (but not necessarily completed). There is a happens-before between the starting stage of the send operation (before) and the completion stage of the receive operation (after). When that HB is satisfied the receive operation is enabled, before that time, it is not enabled, even if it is active.

The "Wait@0" for the send operation may (is permitted to) delay its return until the data has been transmitted ... similar wording.

**Comment by jprotze** [comment]
Ok, so we have at least two HB-relations for MPI operations. 
1. An optimistic or minimal HB-relation. This HB-relation is relevant when reasoning about causality and synchronization semantics to, e.g., avoid data races in RMA communication.
2. An pessimistic or non-minimal HB-relation, that considers worst-case synchronization introduced by a possible implementation and is relevant when reasoning about possible deadlocks (e.g., send-send deadlocks).

So, before using the term HB-relation to define a term like **enable**, it should be clear which HB-relation is used. I think, a proper definition of the two HB-relations would also be great.

Your answer and also [@bosilca 's comment in the PR](https://github.com/mpi-forum/mpi-standard/pull/760#discussion_r1041214583) seems to answer my question: the intention for defining **enabled** is that only events ordered before the operation are relevant for the definition of enabled. 
According to the current definition, leaving aside the colloquial notion, Wait@1 enables Isend@0, if we agree that Isend@0 happens before Wait@1. They are in happens-before relation and are semantically-related procedure calls.
So, this aspect also needs a fix.

**Comment by Wee-Free-Scot** [comment]
(1) is related to enabled (the operation is nonlocal before it is enabled).
(2) is related to progress (the operation is local when it only needs progress).

"Wait@1 enables Isend@0" does not fit with the usage of the word throughout the MPI Standard and is, therefore, "wrong" (although, without a definition, no-one can make such a strong statement).

"Wait@1 enables Isend@0" assumes weak progress and a large message, Wait@1 is sufficient but not necessary to permit the source to transmit the message data into the destination's buffer (progress). However, writing the message data into the destination's buffer is neither sufficient nor necessary for the send operation to complete at the source. Enabled is extracting only what is necessary for the completion stage of the operation. Standard mode send is permitted to delay completion until the matching receive is started (but won't always, in practice). Synchronous mode send is required to delay completion until the matching receive is started (but won't always have sent all the message data when it completes).

Once the Isend@0 and the Irecv@1 calls have both returned, both operations are enabled, meaning their completing now depends only on progress, not (no longer) on a specific semantically-related MPI procedure call. With sufficient progress (provided either by a strong progress implementation, or by any MPI procedure calls that guarantee progress, one of which is the Wait mentioned), the operation will/must complete, whatever (else) the other MPI processes have done.

> the intention for defining enabled is that only events ordered before the operation are relevant for the definition of enabled.

This is problematic -- I'm not sure I agree, but I think I don't understand your intended meaning well enough to suggest an alternative. The operation is not a single moment in time -- "ordered before the operation" is not well-defined; do you mean "ordered before *any*/*all* stage(s) of the operation" or "ordered before a particular stage of the operation" if so, which stage? What do you mean by "event"? Perhaps you mean "MPI procedure call" but you probably mean "specific semantically-related MPI procedure call"?

Fundamentally, we are calling out a fairly subtle state-transition from "completion can be delayed until the operation is enabled, which requires specific things to happen in specific places" to "completion can be delayed until sufficient progress happens, which (for a weak progress implementation) requires unspecific things to happen in specific places"

Enabled removes the veto, but you still need to meet quorum.

**Comment by Wee-Free-Scot** [comment]
Feedback from the reading at the Dec'22 voting meeting:

* The general idea (defined the term enabled) is good.
* The wording (especially "happens-before") is poor.
* Further discussion (in the Semantics WG) is needed.
* A virtual meeting (mid Q1'23) would be useful.

Comments from the meeting chat:

> From Jeff Hammond (NVIDIA) to Everyone 05:02 PM
Happens-before is a well-defined computer science concept.  Why do we want to avoid that and invent new words?
From Jim Dinan to Everyone 05:06 PM
Can we simplify this to: All semantically-related MPI procedure calls have been made at all MPI processes.
From Jim Dinan to Everyone 05:12 PM
I think using happens-before leads to a circular-is definition. It says that all operations that must happen before an operation is enabled have happened.
From Joachim Jenke (RWTH) to Everyone 05:21 PM
enabled: the state where completion only depends on local procedure calls.
From Jim Dinan to Everyone 05:28 PM
Where “local” means by all involved MPI processes. Sorry if this is obvious, took me a moment to read “local” correctly. 🙂
The state where the operation can be completed through local procedure calls at all involved MPI processes.
From Jim Dinan to Everyone 05:49 PM
This is a good example of where we could use "enabled".
From Jim Dinan to Everyone 05:59 PM
Line 2743 was the instance I was thinking of. This paragraph has the same “must complete” description but doesn’t use the term “enabled” (although it should).

**Comment by Wee-Free-Scot** [comment]
This issue has been superseded by issue https://github.com/mpi-forum/mpi-issues/issues/669
